dm/pkg/checker/table_structure.go

```diff
@@ -60,71 +71,106 @@ func (o *incompatibilityOption) String() string {
 type TablesChecker struct {
 	db     *sql.DB
 	dbinfo *dbutil.DBConfig
-	tables map[string][]string // schema => []table; if []table is empty, query tables from db
+	tables []*filter.Table // tableID => filter.Table
 }
 
 // NewTablesChecker returns a RealChecker.
-func NewTablesChecker(db *sql.DB, dbinfo *dbutil.DBConfig, tables map[string][]string) RealChecker {
-	return &TablesChecker{
+func NewTablesChecker(db *sql.DB, dbinfo *dbutil.DBConfig, tables []*filter.Table) RealChecker {
+	c := &TablesChecker{
 		db:     db,
 		dbinfo: dbinfo,
 		tables: tables,
 	}
+	return c
 }
 
 // Check implements RealChecker interface.
-func (c *TablesChecker) Check(ctx context.Context) *Result {
+func (c *TablesChecker) Check(ctx context.Context) (*Result, error) {
 	r := &Result{
 		Name:  c.Name(),
 		Desc:  "check compatibility of table structure",
 		State: StateSuccess,
 		Extra: fmt.Sprintf("address of db instance - %s:%d", c.dbinfo.Host, c.dbinfo.Port),
 	}
 
+	startTime := time.Now()
 	var (
-		err        error
-		options    = make(map[string][]*incompatibilityOption)
-		statements = make(map[string]string)
+		optCh   = make(chan *incompatibilityOption, MaxOptions*len(c.tables))
+		errCh   = make(chan error, 1)
+		checkWg sync.WaitGroup
 	)
-	for schema, tables := range c.tables {
-		if len(tables) == 0 {
-			tables, err = dbutil.GetTables(ctx, c.db, schema)
-			if err != nil {
-				markCheckError(r, err)
-				return r
-			}
-		}
-
+	concurrency := getConcurrency(len(c.tables))
+	maxConnections, err := utils.GetMaxConnections(ctx, c.db)
+	if err != nil {
+		return r, err
+	}
+	if concurrency > maxConnections {
+		concurrency = maxConnections
+	}
+	onethread := len(c.tables) / concurrency
+	defer func() {
+		close(optCh)
+		close(errCh)
+	}()
+	log.L().Logger.Info("start to check tables", zap.Int("concurrency", concurrency), zap.Int("onthread check num", onethread), zap.Int("all table num", len(c.tables)))
+	checkFunc := func(tables []*filter.Table) {
+		defer func() {
+			checkWg.Done()
+		}()
 		for _, table := range tables {
-			tableName := dbutil.TableName(schema, table)
-			statement, err := dbutil.GetCreateTableSQL(ctx, c.db, schema, table)
+			statement, err := dbutil.GetCreateTableSQL(ctx, c.db, table.Schema, table.Name)
 			if err != nil {
 				// continue if table was deleted when checking
 				if isMySQLError(err, mysql.ErrNoSuchTable) {
 					continue
 				}
-				markCheckError(r, err)
-				return r
+				if len(errCh) == 0 {
+					errCh <- err
+				}
+				break
 			}
 
 			opts := c.checkCreateSQL(ctx, statement)
-			if len(opts) > 0 {
-				options[tableName] = opts
-				statements[tableName] = statement
+			for _, opt := range opts {
+				opt.tableID = table.String()
+				optCh <- opt
 			}
 		}
 	}
 
-	for name, opts := range options {
-		if len(opts) == 0 {
-			continue
+	for i := 0; i < concurrency; i++ {
+		if onethread == 0 {
+			checkWg.Add(1)
+			go checkFunc(c.tables[0:])
```

Comment:
when start a new backaround thread , we better to pass ctx in, and use selcet to listen this ctx

this is how golang's way to handle timeout or interrupts